### PR TITLE
[v6r14] new threadpool will be used by the web server

### PR DESCRIPTION
**Pull request by zmathe**

### CODE REVIEW

**Comment by fstagni** [comment]
You should add 'futures' to the 'requirements.txt' file, otherwise the Travis tests will fail.

**Comment by fstagni** [comment]
Question: shouldn't the "old" DIRAC ThreadPool also inherit from the new abstract class `Pool`?

**Comment by chaen** [comment]
and shouldn't this go to v6r15 ?

**Comment by fstagni** [comment]
If it doesn't impact anything else it can also go for v6r14.

**Comment by fstagni** [comment]
Anyway: make the PR to Externals for adding `futures`. It won't be merged immediately but anyway we shouldn't forget about it.

File: Core/Utilities/ThreadPoolExecutor.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+"""
+import sys
+gMissingLibrary = False
+try:
+  import concurrent.futures
+except ImportError:
+  from DIRAC.Core.Utilities.ThreadPool import ThreadPool
+  global gMissingLibrary
+  gMissingLibrary = True
```
Comment (fstagni) [inline Core/Utilities/ThreadPoolExecutor.py]: Please add a "FIXME" that this should be cleaned once `futures` is part of the Externals

File: Core/Utilities/ThreadPoolExecutor.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    if gMissingLibrary:
+      gLogger.info("DIRAC ThreadPool is used!")
+      gThreadPoolExecutor = ThreadPool( 1, 500 )
+      gThreadPoolExecutor.daemonize()
+    else:
+      gThreadPoolExecutor = ThreadPoolExecutor( 500 )
+      gLogger.info("concurrent.futures.ThreadPoolExecutor is used")
+  return gThreadPoolExecutor
+ 
+if __name__ == "__main__":
```
Comment (fstagni) [inline Core/Utilities/ThreadPoolExecutor.py]: Is this necessary?

**Comment by fstagni** [comment]
The test in Travis still fails (with different error message), please check what's missing (probably another library)

File: Core/Utilities/ThreadPoolExecutor.py (hunk 2 above)
Comment (zmathe) [inline Core/Utilities/ThreadPoolExecutor.py]: I used for testing... I deleted as we already have an unit test...

**Comment by zmathe** [comment]
@fstagni The old DIRAC threadPool can be inherited from the Pool, but I do not know what will happen as we will have multiple inheritance... I am going to check it...

File: Core/Utilities/Pool.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  def generateJobAndQueueIt( self,
+                             oCallable,
+                             args = None,
+                             kwargs = None,
+                             sTJId = None,
+                             oCallback = None,
+                             oExceptionCallback = None,
+                             blocking = True ):
+    """
+    :param object oCallable the method which will be executed on a thread.
```
Comment (andresailer) [inline Core/Utilities/Pool.py]: there need to be colons after the parameter names, e.g.
`:param object oCallable: the method...`
also after return, i.e.:
`:return:`

**Comment by fstagni** [comment]
Not for now.

**Pull request closed without merging**